### PR TITLE
Add stateless app-admin registry publishing

### PR DESCRIPTION
**Pull request by mzywang**
## Summary
- add deterministic create/finalize publishing endpoints for app administrators without a durable session ledger
- derive resume and publication state directly from staging objects and the strongly read registry index
- wire create-only signed GCS uploads and idempotent finalization into the compact registry writer

## Test plan
- [x] `go run github.com/golangci/golangci-lint/v2/cmd/golangci-lint@v2.12.1 run`
- [x] `go vet ./...`
- [x] `go test -timeout=15m ./...`
- [x] `go build ./...`

## Stack
- Depends on #3105

Made with [Cursor](https://cursor.com)

### CODE REVIEW

File: gestaltd/internal/appregistry/publish_gcs.go [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+	uploadURL, err := s.signedURL(client, bucket, object, &storage.SignedURLOptions{
+		Scheme:  storage.SigningSchemeV4,
+		Method:  "PUT",
+		Headers: signedHeaders,
+		Expires: expiresAt.UTC(),
+	})
+	if err != nil {
+		return SignCreateUploadResult{}, fmt.Errorf("sign upload URL for %s: %w", storageURL, err)
+	}
+	return SignCreateUploadResult{UploadURL: uploadURL, ExpiresAt: expiresAt.UTC(), Headers: headers}, nil
```
[the remark this answers is not in the document]
Comment (mzywang) [reply gestaltd/internal/appregistry/publish_gcs.go]: Fixed on the current branch. Staging uploads no longer sign source-ref metadata; the exact returned header set is now regression-tested against the signed header set.

File: gestaltd/internal/appregistry/publish_gcs.go [hunk 2]
```diff
@@ -0,0 +1,597 @@
...
+	}
+	var missing []string
+	for permission := range required {
+		if _, ok := granted[permission]; !ok {
+			missing = append(missing, permission)
+		}
+	}
+	if len(missing) > 0 {
+		return fmt.Errorf("gcs registry permissions missing: %s", strings.Join(missing, ", "))
+	}
```
[the remark this answers is not in the document]
Comment (mzywang) [reply gestaltd/internal/appregistry/publish_gcs.go:606]: Clarified on the current branch. We intentionally retain storage.objects.delete because GCS requires that IAM permission to overwrite an existing object, including generation-matched CAS replacements of index.json and retention.json. The publish path never issues a Delete API call. I removed the unnecessary storage.objects.update permission and the bucket metadata probe/storage.buckets.get dependency, and added an exact permission-set test.

File: gestaltd/internal/server/handlers_app_admin_registry_publish.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+	if err != nil {
+		writeError(w, http.StatusServiceUnavailable, "app registry is unavailable")
+		return nil, "", "", false
+	}
+	publicRoot, err := registry.PublicURL()
+	if err != nil {
+		writeError(w, http.StatusServiceUnavailable, "app registry is unavailable")
+		return nil, "", "", false
+	}
+	return s.appRegistryPublish, storageRoot, publicRoot, true
```
[the remark this answers is not in the document]
Comment (mzywang) [reply gestaltd/internal/server/handlers_app_admin_registry_publish.go]: Fixed on the current branch. The service is bound to one writable registry and both Begin and Finalize reject apps enrolled in another registry; handler and service regression tests cover this boundary.

File: gestaltd/internal/appregistry/publish_admin.go [hunk 4]
```diff
@@ -0,0 +1,668 @@
...
+		digest := strings.ToLower(strings.TrimSpace(artifact.SHA256))
+		if platform == "" || filename == "" || digest == "" {
+			return fmt.Errorf("%w: artifact platform, filename, and sha256 are required", ErrPublishDeclarationInvalid)
+		}
+		if len(digest) != 64 {
+			return fmt.Errorf("%w: artifact %q sha256 must be 64 hex characters", ErrPublishDeclarationInvalid, platform)
+		}
+		if limits.MaxArtifactBytes > 0 && artifact.Size > limits.MaxArtifactBytes {
+			return fmt.Errorf("%w: artifact %q exceeds size limit", ErrPublishArtifactLimit, platform)
+		}
```
[the remark this answers is not in the document]
Comment (mzywang) [reply gestaltd/internal/appregistry/publish_admin.go]: Fixed on the current branch. Artifact size must be positive during declaration validation, and omitted/zero sizes are covered at the service and HTTP boundary as typed 400 responses.

File: gestaltd/internal/appregistry/writer_gcloud.go [hunk 5]
```diff
@@ -86,6 +86,27 @@ func (s *GcloudObjectStore) WriteCatalogObject(input WriteCatalogObjectInput) er
...
+		"--if-generation-match=0",
+		"--custom-metadata=" + metadata,
+	}
+	if input.SourceGeneration > 0 {
+		args = append(args, "--source-generation="+strconv.FormatInt(input.SourceGeneration, 10))
+	}
+	args = append(args, input.SourceURL, input.DestURL)
+	_, err := s.Runner.Run("gcloud", args...)
+	return err
+}
```
[the remark this answers is not in the document]
Comment (mzywang) [reply gestaltd/internal/appregistry/writer_gcloud.go]: Fixed structurally. Promotion is now a separate RegistryObjectPromoter capability used only by stateless finalization; GcloudObjectStore implements only the base writer store and its weaker PromoteObject method was removed. The stateless service requires a single composite store/promoter backend.

File: gestaltd/internal/appregistry/publish_admin.go [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+	defer func() { _ = os.Remove(manifest.EntryObject.LocalPath) }()
+
+	req := PublishRequest{Manifest: manifest, SourceRef: sourceRef}
+	if err := s.Writer.Preflight(req, PublishProgress{}); err != nil {
+		return nil, err
+	}
+	result, err := s.publishWithCatalogRetry(req)
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (mzywang) [reply gestaltd/internal/appregistry/publish_admin.go:161]: Fixed in the base writer and restacked here. Equivalent entries are reconciled if the winner appears either between the two reads or at the create-only write; conflicting/malformed winners fail closed. The same reconciliation recognizes typed storage preconditions and raw gcloud 412 errors. Deterministic tests cover both race windows.

File: gestaltd/internal/appregistry/publish_errors.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+		return 400
+	case errors.Is(err, ErrPublishVersionConflict),
+		errors.Is(err, ErrPublishUploadMismatch),
+		errors.Is(err, ErrPublishReconcileMismatch):
+		return 409
+	case errors.Is(err, ErrPublishRegistryNotEnrolled):
+		return 404
+	default:
+		return 502
+	}
```
[the remark this answers is not in the document]
Comment (mzywang) [reply gestaltd/internal/appregistry/publish_errors.go:40]: Fixed. Wrapped ErrRegistryEntryConflict and ErrObjectPreconditionFailed now map to HTTP 409; catalog generation conflicts are retried internally first, while genuine backend errors remain 502. Table tests cover exact and wrapped errors.

File: gestaltd/internal/appregistry/publish_admin.go [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+	manifest, err := s.buildFinalManifest(input, canonical, publishID, digest, version, publishedAt)
+	if err != nil {
+		return nil, err
+	}
+	defer func() { _ = os.Remove(manifest.EntryObject.LocalPath) }()
+
+	req := PublishRequest{Manifest: manifest, SourceRef: sourceRef}
+	if err := s.Writer.Preflight(req, PublishProgress{}); err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (mzywang) [reply gestaltd/internal/appregistry/publish_admin.go:154]: Fixed. Finalize now builds the manifest and runs Writer.Preflight before any promotion. A conflicting orphan entry therefore causes zero promotion calls and creates no final artifact; matching pre-promoted artifacts remain resumable. Both paths have regression coverage.

File: gestaltd/internal/appregistry/publish_declaration_identity.go [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+			return fmt.Errorf("%w: artifact %q size must be greater than zero", ErrPublishDeclarationInvalid, platform)
+		}
+		if limits.MaxArtifactBytes > 0 && artifact.Size > limits.MaxArtifactBytes {
+			return fmt.Errorf("%w: artifact %q exceeds size limit", ErrPublishArtifactLimit, platform)
+		}
+		if _, ok := platforms[platform]; ok {
+			return fmt.Errorf("%w: duplicate platform %q", ErrPublishDeclarationInvalid, platform)
+		}
+		platforms[platform] = struct{}{}
+	}
```
[the remark this answers is not in the document]
Comment (mzywang) [reply gestaltd/internal/appregistry/publish_declaration_identity.go:133]: Fixed. Declaration validation now rejects duplicate canonical filenames across platforms before any upload is signed, returning the existing typed declaration error/HTTP 400. Tests cover distinct filenames, duplicate platforms, duplicate filenames, and whitespace-normalized duplicates.

**Comment by mzywang** [comment]
Closing at the author's request. The stateless API will be reimplemented with a smaller footprint and explicit per-app pilot gating.
